File: University Projects/Java Projects/Design and Programming Environments/Proiect/Persistence/src/main/java/app/persistance/repository/DBUtils.java
package app.persistance.repository;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.util.Properties;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

public class DBUtils {
    private final Properties dbProps;

    private Connection instance;

    private static final Logger logger = LogManager.getLogger();

    public DBUtils(Properties dbProps) {
        this.dbProps = dbProps;
    }

    private Connection createNewConnection() {
        logger.traceEntry();

        String url = dbProps.getProperty("URL");
        logger.info("Trying to connect to database...: {}", url);
        Connection connection = null;
        try {
            connection = DriverManager.getConnection(url);
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error getting connection " + e);
        }
        return connection;
    }

    public Connection getConnection() {
        logger.traceEntry();
        try {
            if (instance == null || instance.isClosed()) {
                instance = createNewConnection();
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error DB " + e);
        }
        logger.traceExit(instance);
        return instance;
    }

    public void closeConnection() {
        logger.traceEntry();
        try {
            if (instance != null && !instance.isClosed()) {
                instance.close();
                logger.info("Connection closed.");
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error closing connection " + e);
        }
        logger.traceExit();
    }
}

File: University Projects/C# Projects/Design and Programming Environments/Proiect/persistence/DBUtils.cs
using Microsoft.Data.Sqlite;
using Serilog;
using System.Configuration;
using System;
using System.Collections.Generic;
using System.Data;
using System.Text;

namespace persistence
{
    public class DBUtils
    {
        private readonly string _connectionString;
        private IDbConnection _instance;
        private readonly ILogger _logger = Log.ForContext<DBUtils>();

        public DBUtils(string connectionString)
        {
            _connectionString = connectionString;
        }

        private IDbConnection CreateNewConnection()
        {
            _logger.Verbose("Entering CreateNewConnection");
            _logger.Information("Trying to connect to database...: {Url}", _connectionString);

            IDbConnection connection = null;
            try
            {
                connection = new SqliteConnection(_connectionString);
                connection.Open();
            }
            catch (Exception e)
            {
                _logger.Error(e, "Error getting connection");
                Console.WriteLine("Error getting connection " + e.Message);
            }
            return connection;
        }

        public IDbConnection GetConnection()
        {
            _logger.Verbose("Entering GetConnection");
            try
            {
                if (_instance == null || _instance.State == ConnectionState.Closed)
                {
                    _instance = CreateNewConnection();
                }
            }
            catch (Exception e)
            {
                _logger.Error(e, "Error DB");
                Console.WriteLine("Error DB " + e.Message);
            }

            _logger.Verbose("Exiting GetConnection");
            return _instance;
        }

        public void CloseConnection()
        {
            _logger.Verbose("Entering CloseConnection");
            if (_instance != null && _instance.State != ConnectionState.Closed)
            {
                try
                {
                    _instance.Close();
                    _logger.Information("Database connection closed successfully.");
                }
                catch (Exception e)
                {
                    _logger.Error(e, "Error closing database connection");
                    Console.WriteLine("Error closing connection " + e.Message);
                }
            }
        }
    }
}
